_posts/2015-08-07-symbolic-operators.md

```diff
@@ -0,0 +1,183 @@
+However, `semigroupSyntax` requires an implicit `Semigroup[A]` value to be in scope.
+Do we have a `Semigroup[Int]` in scope?
+
+Yes we do. Our import also provides an implicit value [`intGroup`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/std/src/main/scala/cats/std/anyval.scala#L23)
+of type `AdditiveCommutativeGroup[Int]`. Leaving aside what *additive*, *commutative*,
+and *group* mean here, this is a subtype of `Semigroup[Int]`, so it matches.
+
+Let's continue with our current example. At this point we have gone form:
+
+```scala
+19 |+| 20 // produces 39
+```
+
+to:
+
+```scala
+semigroupSyntax[Int](19)(intGroup) |+| 20
+```
+
+But we aren't out of the woods yet! We still need to see how this expression
+is evaluated.
+
+Of macros and machinists
+------------------------
+
+Looking at how the `|+|` method is
+[implemented](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/core/src/main/scala/cats/syntax/semigroup.scala#L13)
+reveals the cryptic `macro Ops.binop[A, A]`. What is this?
+
+Following the rabbit hole farther, we come to
+[`cats.macros.Ops`](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/macros/src/main/scala/cats/macros/Ops.scala#L6)
+which provides the macro implementation that `|+|` is using. Aside from a
+[suggestively named](https://github.com/non/cats/blob/82dbf4076572dfbb6e29dd49875f5e5d929f80be/macros/src/main/scala/cats/macros/Ops.scala#L18)
+item in the `operatorNames` map, we don't have any clues what is going on.
+
+The [machinist](https://github.com/typelevel/machinist) project was created
+to optimize exactly this kind of implicit syntax problem. What will happen here
+is that `operatorNames` describes how to rewrite expressions using type
+classes. Long-story short, we will transform:
+
+```scala
+semigroupSyntax[Int](19)(intGroup) |+| 20
+```
+
+into:
+
+```scala
+intGroup.combine(19, 20)
+```
+
+The suggestive map item tells us that we should rewrite the `|+|` operator
+to method calls on the given type class (i.e. `intGroup`) using `.combine`.
+
+Finishing up
+------------
+
+Just to confirm that we're done, let's look at what `intGroup.combine` will do.
+We started with a call to `AdditiveCommutativeGroup[Int]`, which will find
+[`intAlgebra`](https://github.com/non/algebra/blob/v0.3.1/std/shared/src/main/scala/algebra/std/int.scala#L12).
+Then we call the [`.additive`](https://github.com/non/algebra/blob/v0.3.1/core/src/main/scala/algebra/ring/Additive.scala#L93)
+method on it to produce a `CommutativeGroup[Int]`.
+
+So putting that together, we can see that calling `intGroup.combine(19, 20)`
+will call `intAlgebra.plus(19, 20)`, and that this is defined as `19 + 20`,
+as we would expect.
+
+Whew!
+
+Conclusion
+----------
+
+This is a lot of machinery. The incredibly terse and expressive syntax it
+enables is quite nice, but you can see that even leaving out one import
+will cause the whole edifice to come tumbling down.
```

Comment:
Yes, I would say that unfortunately in Scala this is the best choice.  We should still say what the shortcomings of the best choice are, lest someone think everything is hunky dory.